### PR TITLE
Add node elevations from DEM files

### PR DESCRIPTION
**Pull request by kylebarron**
Refactor of #343 to use [`demquery`](https://github.com/kylebarron/demquery).

**Overview**:

- I split my code into a new Python package on PyPI. 
- Since learning about [Virtual Rasters](https://gdal.org/drivers/raster/vrt.html), the code to find elevations is an order of magnitude simpler. I don't do any joining of nearby rasters; GDAL does all of that.
- I discovered that `interp2d` gives those warnings only when the grid is not the expected size for the type of interpolation: 3x3 for linear interpolation, 5x5 for cubic interpolation, and 7x7 for quintic interpolation. So I removed the option to manually choose the size of the interpolation region. If `interp_kind` is `linear`, it creates a 3x3 grid.
- Checks for nodata values; the virtual raster can have regions with no data that are inside the file's bounds, if the provided files didn't create an exact rectangle. My package raises an error if it sees those values in the interpolation region.
- I'm working on writing tests in the demquery package. ~~I'm having issues with rasterio working on Travis CI~~ the osmnx `.travis.yml` was quite helpful.
- Do you have plans to deprecate Python 2.7 and 3.4? Python 3.4 is already end of life, and 2.7 is in a month. I made `demquery` 3.5+ compatible, but haven't tested Python 2.7.
- Right now, `demquery` is only on PyPI, and I think that's why the current Travis build is failing.

**Example**:

Download both the files 
[`USGS_NED_13_n33w117_IMG`](https://prd-tnm.s3.amazonaws.com/StagedProducts/Elevation/13/IMG/USGS_NED_13_n33w117_IMG.zip) 
[`USGS_NED_13_n34w117_IMG`](https://prd-tnm.s3.amazonaws.com/StagedProducts/Elevation/13/IMG/USGS_NED_13_n34w117_IMG.zip) 
450MB extracted each

```py
import osmnx as ox
west = -116.66587829589844
south = 32.90120475753773
north = 33.094201303793194
east = -116.46263122558594
G = ox.graph_from_bbox(north=north, south=south, west=west, east=east)
dem_paths = ['USGS_NED_13_n33w117_IMG.img', 'USGS_NED_13_n34w117_IMG.img']
# interp_kind can be None, linear, cubic, or quintic
G = add_node_elevations_from_dem(G, dem_paths, interp_kind=None)
G = add_edge_grades(G)
```

### CODE REVIEW

**Comment by gboeing** [comment]
`demquery` needs to be live on conda-forge for the CI to run successfully

**Comment by gboeing** [comment]
> Do you have plans to deprecate Python 2.7 and 3.4? Python 3.4 is already end of life, and 2.7 is in a month. I made demquery 3.5+ compatible, but haven't tested Python 2.7.

OSMnx is dropping official py2.x and py3.4 support in the next release.

File: osmnx/elevation.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 # Web: https://github.com/gboeing/osmnx
 ################################################################################
 
+import demquery
```
Comment (gboeing) [inline osmnx/elevation.py]: This should be handled as an optional import. For example see how scikit-learn components are imported here: https://github.com/gboeing/osmnx/blob/bde270d4a148d8f8911aef59c86a099adc76c24f/osmnx/geo_utils.py#L39-L42
And then checked/handled here: https://github.com/gboeing/osmnx/blob/bde270d4a148d8f8911aef59c86a099adc76c24f/osmnx/geo_utils.py#L427-L428

File: requirements.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
+demquery>=0.2.0
```
Comment (gboeing) [inline requirements.txt]: Place this dependency in `requirements-dev.txt` instead, as this is an optional dependency for non-core functionality (see previous comment above).

File: osmnx/elevation.py (hunk 1 above)
Comment (kylebarron) [reply osmnx/elevation.py]: Ah makes sense. I wasn't sure how to do optional imports but I agree it should be one

**Comment by kylebarron** [comment]
> > Do you have plans to deprecate Python 2.7 and 3.4? Python 3.4 is already end of life, and 2.7 is in a month. I made demquery 3.5+ compatible, but haven't tested Python 2.7.
> 
> OSMnx is dropping official py2.x and py3.4 support in the next release.

Great. I just asked because as of now [`setup.py`](https://github.com/gboeing/osmnx/blob/master/setup.py) still includes 2.x and 3.4.

**Comment by kylebarron** [comment]
I tried uploading to conda-forge but I'm getting uninformative build errors https://github.com/conda-forge/staged-recipes/pull/10355. I'll have to take another stab at this tomorrow.

**Comment by kylebarron** [comment]
Checks now pass on my [conda-forge recipe](https://github.com/conda-forge/staged-recipes/pull/10355), so ideally it'll be on conda-forge soon.

I'll open a new PR for adding elevations to edges?

**Comment by kylebarron** [comment]
Just toggling to re-run travis

**Comment by kylebarron** [comment]
Ok now the feedstock is rebuilt and the [package page is live](https://anaconda.org/conda-forge/demquery) so I'll toggle it once more.

**Comment by kylebarron** [comment]
Do you want a test on this, so that the coverage percentage doesn't go down? I'm not sure what the easiest way to test is? Find a lightweight DEM and then download it every time?

**Comment by gboeing** [comment]
@kylebarron what's the status on this PR? Were you able to identify a lightweight DEM file and incorporate it into the tests to cover the new functionality?

**Comment by gboeing** [comment]
Closed for inactivity. Can be reopened in the future as needed.

**Comment by kylebarron** [comment]
> Were you able to identify a lightweight DEM file and incorporate it into the tests to cover the new functionality?

I never got a reply to my questions, so I never did anything further. 

Small DEM files are easy to find, e.g. `https://raw.githubusercontent.com/visgl/deck.gl-data/master/website/terrain.png` of SF:


I think adding a test should be pretty simple. I guess the question is how to have a reproducible network of SF in the test. I haven't looked into your test code recently.

**Comment by janiksudo** [comment]
I would like to revive this issue(and help with it) , and add DEM files as another source of elevation next to Google Maps and Airmap.

What needs to be done to conclude your work @kylebarron, and what needs to be done in your opinion @gboeing - I guess tests to cover the new functionality, anything else?

**Comment by kylebarron** [comment]
As of my last commit to this branch, it was working correctly and just needed test cases. I haven't kept up with any changes in `osmnx` since this PR

**Comment by gboeing** [comment]
@janikmu been a while since I looked through this PR, but off the top of my head:

  1. lightweight but thorough tests
  2. double-checking functionality with rest of codebase as there have been some changes in recent months
  3. some sort of ongoing maintenance plan

The first two are straightforward, but I'm unsure of the latter, which is increasingly important as OSMnx accumulates features beyond its core functionality (and which I do not have personal use cases for). For example, this PR adds `demquery` as an optional dependency, but it looks like this package is [no longer](https://github.com/conda-forge/demquery-feedstock/pull/1) being kept up to date on conda-forge as it has a stale auto-PR from January.

This is not meant as a swipe against @kylebarron's `demquery` project in any way: it looks like a useful tool! But, to add a dependency, we also need to have some assurance of near-permanent support for the dependency package itself as well as its conda-forge distribution. The reason is simple: I'm the primary maintainer and nearly sole tech support, and I separately already work 70+ hours a week. When a new feature is contributed, and the contributor eventually drifts away from using OSMnx, I'm on the hook for any additional marginal ongoing maintenance of the codebase and keeping an eye on any dependencies that don't have lots and lots of contributors and ongoing commits.

So the short story is: I'm open to this PR, but I also want you to be aware of these caveats regarding ongoing maintenance and time burdens. I'm open to any comments that may put them at rest.

**Comment by kylebarron** [comment]
Thanks, I must've missed the conda-forge PR before; I previously only published to pip, and conda-forge packaging is much more difficult than publishing to pypi.

In any case, the internals of `demquery` is a simple, well-documented [300-line script](https://github.com/kylebarron/demquery/blob/master/demquery/demquery.py). @janikmu could copy the file if you didn't want to link to an external dependency.

**Comment by janikmu** [comment]
Good comment, @gboeing ! I really did not see it from this side. Maybe it then would be better to not bloat osmnx with more non-core functionality and do elevation from DEM files - > networkx somewhere else. (As elevation really is only another node attribute in the networkx graph, and can be added after the network is created by osmnx).

Somewhere else could mean at the user-level (I, for my use-case can do it following the code from @kylebarron ) or maybe I will even write a module for it if I have the time.

On the other hand, if `demquery` is a no-brainer dependency that needs no maintenance, this is your call @gboeing to decide, if it is outside of core functionality or could be helpful.

**Comment by gboeing** [comment]
Thanks all for the comments. At this point I'm inclined to leave DEM elevation out of OSMnx itself. It looks like it would be relatively straightforward for a user to use demquery to add node elevations after building a graph.

If you do write a short general script adapting the code here to load graph node elevations using demquery, it'd be awesome to include it as an example at https://github.com/gboeing/osmnx-examples for others to reuse. That could be a great option, if you're so inclined.

**Comment by kylebarron** [comment]
> If you do write a short general script adapting the code here to load graph node elevations using demquery, it'd be awesome to include it as an example at [gboeing/osmnx-examples](https://github.com/gboeing/osmnx-examples) for others to reuse. 

I think that's a great suggestion if @janikmu is interested. You can copy the function from this PR into a notebook and provide it as an example for adding elevation nodes.

**Pull request closed without merging**